llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -9105,7 +9105,7 @@ getRegistersForValue(SelectionDAG &DAG, const SDLoc &DL,
   // Get the actual register value type.  This is important, because the user
   // may have asked for (e.g.) the AX register in i32 type.  We need to
   // remember that AX is actually i16 to get the right extension.
-  const MVT RegVT = *TRI.legalclasstypes_begin(*RC);
+  MVT RegVT = *TRI.legalclasstypes_begin(*RC);
```

Comment:
The generic code change feels wrong. It feels really wrong that this just takes the first element of RC. Mutating this later just makes it harder to follow. I think this block of code should either not depend on RegVT at all, or this needs to be a loop over all legal class types

Reply:
Did you mean that we should keep the RegVT constant, but choose not just the first legal value type for creating the SelectionDAG? Such fix is like using a lambda function with `for loop` for choosing the reasonable value type, in this case, not fp16 but bf16.

Reply:
Something like that

Reply:
Adding a new hook that implements TRI.legalclasstypes_begin is not what I meant. That doesn't really address the issue, and just forces more complexity into every target. I meant all of this code here can be factored into a function and repeated for every type in legalclasstypes until one works